### DISCUSSION TITLE
UMP

### DISCUSSION THREAD

**Discussion by SnailRabbids** [Poing-Studios/Godot-AdMob-Android-iOS#123] [Q&A] [answered]
Do I need to do anything after asking for user consent? I mean, if the user selects "Do not consent", can I show them ads or not? As I uderstand if the user doesn't agree, the ads just won't be personalized, but I don't need to change my code because of that, right?

So my steps are: initialize the plugin -> ask for his consent -> here I don't care what the user chose, I can show my ads with show_banner() etc.

Thanks :)

**Comment by SnailRabbids** [comment]
@gumaciel is it ok, that I can request user consent few times?
click on "REQUEST USER CONSENT" -> shows GDPR window -> accept/decline/close -> click again -> shows again -> ...
Is there a way to check if user consent has been shown? Or it means, that Consent Status is still "Required" for some reason?
I also noticed that there is always a "Request Consent from European Users Form **_dismissed_**" even if I click "Consent":(

![image](https://user-images.githubusercontent.com/36141552/183066337-d3f78a0e-c5f7-4aaa-bcdf-b0d5f41ceafe.png)

**Comment by gumaciel** [comment, marked as answer]
Hi @SnailRabbids , answering your questions:
> I mean, if the user selects "Do not consent", can I show them ads or not

Ads will still appear, you may not be able to see them as Google has no ads available for you at the moment.

Note that when you mark that you do not consent to show personalized ads, the amount of available ads is reduced, when you are testing unfortunately it is common not to appear, because from what I understand Google does not provide "Non-customized" ads for testing, which I I believe it is their mistake.

>  As I uderstand if the user doesn't agree, the ads just won't be personalized, but I don't need to change my code because of that, right?

Yes, you are right, you don't need to change your code

> So my steps are: initialize the plugin -> ask for his consent -> here I don't care what the user chose, I can show my ads with show_banner() etc.

That's correct

> is it ok, that I can request user consent few times?

Yes, even the "reset user consent" option has been added to do this: https://github.com/Poing-Studios/godot-admob-android/issues/57

> Is there a way to check if user consent has been shown?

If you press "REQUEST USER CONSENT", you will notice that you will receive a message: "User consent obtained", that means that the consent form was shown for the user and he accepted OR not (there is no way to know if he fact accepted or denied, only if it was obtained)

> Or it means, that Consent Status is still "Required" for some reason?

I didn't understand it very well, but I don't think so, as I see it, the Status is usually just for the developer to get feedback from the user iteration

Note that UMP is not mandatory in all countries.

`Under the Google EU User Consent Policy, you must make certain disclosures to your users in the European Economic Area (EEA) along with the UK and obtain their consent to use cookies or other local storage, where legally required, and to use personal data (such as AdID) to serve ads. This policy reflects the requirements of the EU ePrivacy Directive and the General Data Protection Regulation (GDPR).`

https://developers.google.com/admob/ump/ios/quick-start#introduction

> I also noticed that there is always a "Request Consent from European Users Form dismissed" even if I click "Consent":(

If the form is appearing and then disappears, then it is correct for me. This message refers to the form was dismissed

---
Read more here: 
https://github.com/Poing-Studios/godot-admob-ios/issues/17
https://groups.google.com/g/google-admob-ads-sdk/c/xfLEdy06pyc

**Reply by SnailRabbids** [reply]
@gumaciel thanks for ur answer :) Just a few clarifications:

> Yes, even the "reset user consent" option has been added to do this: https://github.com/Poing-Studios/godot-admob-android/issues/57

I mean, I can request user consent few times w/o reset. Just click on "REQUEST USER CONSENT", then again and again.

> Or it means, that Consent Status is still "Required" for some reason?

Relates to first question, every time I click on "REQUEST USER CONSENT" button I see the form, because status is "REQUIRED", but should be "OBTAINED", no?
https://github.com/Poing-Studios/godot-admob-android/blob/master/admob/src/main/java/com/poing/admob/AdMob.java#L211

**Reply by gumaciel** [reply]
> I mean, I can request user consent few times w/o reset. Just click on "REQUEST USER CONSENT", then again and again.

As far as I know, the form will only appear again if it is reset and the Plugin doesn't currently do this, but you can change it here:

https://github.com/Poing-Studios/godot-admob-editor/blob/master/addons/admob/src/singletons/MobileAds.gd#L48-L50

Change this function to: 

```gdscript
func request_user_consent() -> void:
	if _plugin:
		_plugin.reset_consent_state()
		_plugin.request_user_consent()
```

If you find a smarter approach and find it useful for everyone I ask you to please open a PR for everyone to take advantage of this improvement

> Relates to first question, every time I click on "REQUEST USER CONSENT" button I see the form, because status is "REQUIRED", but should be "OBTAINED", no?

I don't believe so, I got these statuses from here while I was developing the Plugin:

https://developers.google.com/admob/ump/android/quick-start#present_the_form_if_required

- **`ConsentStatus.UNKNOWN`**: Unknown consent status.
- **`ConsentStatus.REQUIRED`**: User consent required but not yet obtained.
- **`ConsentStatus.NOT_REQUIRED`**: User consent not required. For example, the user is not in the EEA or the UK.
- **`ConsentStatus.OBTAINED`**: User consent obtained. Personalization not defined.

In other words: If the form is REQUIRED, the form must appear.

If the form was OBTAINED, it means that the consent form was closed and the user consented or not to the personalization of ads.

**Reply by gumaciel** [reply]
Wait, are you saying that whenever you click the REQUEST USER CONSENT button the form appears? I believe this could be a bug, as the form should only appear again if RESET USER CONSENT is asked

**Reply by SnailRabbids** [reply]
exactly!

**Reply by SnailRabbids** [reply]
Can I ask you to try it by yourself? (Android)

**Reply by gumaciel** [reply]
Sorry but i can't do it right now, seems like Google is with problem on their servers: 
![image](https://user-images.githubusercontent.com/20030153/183101374-a80a8edd-6403-40f0-ab67-af5fc9bc22b9.png)

**Reply by SnailRabbids** [reply]
@gumaciel are you sure you have enabled your GDPR on admob.com? I had same error when it was disabled.

**Reply by SnailRabbids** [reply]
because form appears for me right now. With this bug, but appears ...

**Reply by SnailRabbids** [reply]
and one more thing, I'm not 100% sure, but yesterday this button worked as expected. Mb google servers are broken for real. I think we could wait for 1-2 days and try again?

**Reply by gumaciel** [reply]
Yes, of course, this weekend i won't be able to work on the Plugin, until this i invite you to join on our Discord server, if you have it: https://discord.gg/YEPvYjSSMk

**Comment by SnailRabbids** [comment]
As you said, now it works fine, thanks you :)

**Comment by kyoz** [comment]
Hello @gumaciel , just try to implement UMP cause now it a NEED from admob.

It's seem the current version of plugin work well but I do face some issues which I don't know is it a normal behavior or an error.

No matter what I choose, I only get this message: "User consent obtained. Personalization not defined."

If I choose "Manage options" -> "Confirm choices". Then there will be no ads at all. It's like you explained above. Maybe non-personalize ad not work with test ads. But the problem is, when I try to show a reward ads. There is no error or anything happen. So it's no way to let user know that they need to re-config the consent in setting.

Do you have any experience on this? Thank you in advanced.

**Comment by kyoz** [comment]
Oh, I just tested it on iOS. when user choose not consent. When trying to show reward ads. It will throw `rewarded_ad_failed_to_show`. But only on iOS, on Android it will do nothing.

**Reply by gumaciel** [reply]
Usually when the user doesn't consent, test ads don't show up, but real ads should show up

On Android it should also throw this error, could you please create an Issue of this bug with the description in the repository?

https://github.com/Poing-Studios/godot-admob-android

**Reply by kyoz** [reply]
Yes, sure @gumaciel. Let me test it a little bit, i'll create a issue for Android soon.
